From ce953f011356257c1e1862d0353f785453559028 Mon Sep 17 00:00:00 2001
From: Dyami Caliri <dyami@dragonframe.com>
Date: Tue, 16 Sep 2014 17:32:22 -0700
Subject: [PATCH] OS X: Handle Qt::AA_MacDontSwapCtrlAndMeta for regular key
 events

The flag Qt::AA_MacDontSwapCtrlAndMeta was only being used with shortcut
detection. Now trying to incorporate flag in normal key events.

Task-number: QTBUG-22775
Change-Id: I9edf97d0de93aa225435c9802372e33641cdac73
===

20160615: Qt 5.6.1 introduces something (_q_dontOverrideCtrlLMB/QT_MAC_DONT_OVERRIDE_CTRL_LMB)
which may make the 1st hunk redundant.

src/plugins/platforms/cocoa/qnsview.mm | 16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

diff --git a/cocoa/qnsview.mm b/cocoa/qnsview.mm
index a63bc4d..14f82e1 100644
--- a/cocoa/qnsview.mm
+++ b/cocoa/qnsview.mm
@@ -1489,15 +1489,16 @@ static QTabletEvent::TabletDevice wacomTabletDevice(NSEvent *theEvent)
 
 + (Qt::KeyboardModifiers) convertKeyModifiers : (ulong)modifierFlags
 {
+    const bool swapCtrlAndMeta = !qApp->testAttribute(Qt::AA_MacDontSwapCtrlAndMeta);
     Qt::KeyboardModifiers qtMods =Qt::NoModifier;
     if (modifierFlags &  NSShiftKeyMask)
         qtMods |= Qt::ShiftModifier;
     if (modifierFlags & NSControlKeyMask)
-        qtMods |= Qt::MetaModifier;
+        qtMods |= (swapCtrlAndMeta ? Qt::MetaModifier : Qt::ControlModifier);
     if (modifierFlags & NSAlternateKeyMask)
         qtMods |= Qt::AltModifier;
     if (modifierFlags & NSCommandKeyMask)
-        qtMods |= Qt::ControlModifier;
+        qtMods |= (swapCtrlAndMeta ? Qt::ControlModifier : Qt::MetaModifier);
     if (modifierFlags & NSNumericPadKeyMask)
         qtMods |= Qt::KeypadModifier;
     return qtMods;
@@ -1521,6 +1522,7 @@ static QTabletEvent::TabletDevice wacomTabletDevice(NSEvent *theEvent)
     quint32 nativeScanCode = 1;
     quint32 nativeVirtualKey = [nsevent keyCode];
 
+    const bool swapCtrlAndMeta = !qApp->testAttribute(Qt::AA_MacDontSwapCtrlAndMeta);
     QChar ch = QChar::ReplacementCharacter;
     int keyCode = Qt::Key_unknown;
 
@@ -1530,7 +1532,7 @@ static QTabletEvent::TabletDevice wacomTabletDevice(NSEvent *theEvent)
     // ALT+E to be used as a shortcut with an English keyboard even though
     // pressing ALT+E will give a dead key while doing normal text input.
     if ([characters length] != 0 || [charactersIgnoringModifiers length] != 0) {
-        if (((modifiers & Qt::MetaModifier) || (modifiers & Qt::AltModifier)) && ([charactersIgnoringModifiers length] != 0))
+        if (((modifiers & (swapCtrlAndMeta ? Qt::MetaModifier : Qt::ControlModifier)) || (modifiers & Qt::AltModifier)) && ([charactersIgnoringModifiers length] != 0))
             ch = QChar([charactersIgnoringModifiers characterAtIndex:0]);
         else if ([characters length] != 0)
             ch = QChar([characters characterAtIndex:0]);
@@ -1661,6 +1663,7 @@ static QTabletEvent::TabletDevice wacomTabletDevice(NSEvent *theEvent)
     ulong lastKnownModifiers = m_lastKnownModifiers;
     ulong delta = lastKnownModifiers ^ modifiers;
     m_lastKnownModifiers = modifiers;
+    const bool swapCtrlAndMeta = !qApp->testAttribute(Qt::AA_MacDontSwapCtrlAndMeta);
 
     struct qt_mac_enum_mapper
     {
@@ -1669,8 +1672,8 @@ static QTabletEvent::TabletDevice wacomTabletDevice(NSEvent *theEvent)
     };
     static qt_mac_enum_mapper modifier_key_symbols[] = {
         { NSShiftKeyMask, Qt::Key_Shift },
-        { NSControlKeyMask, Qt::Key_Meta },
-        { NSCommandKeyMask, Qt::Key_Control },
+        { NSControlKeyMask, (swapCtrlAndMeta ? Qt::Key_Meta :  Qt::Key_Control) },
+        { NSCommandKeyMask, (swapCtrlAndMeta ? Qt::Key_Control : Qt::Key_Meta) },
         { NSAlternateKeyMask, Qt::Key_Alt },
         { NSAlphaShiftKeyMask, Qt::Key_CapsLock },
         { 0ul, Qt::Key_unknown } };
